### PR TITLE
fix(svelte): infer props for generic Svelte components

### LINKED ISSUE #14669: 🐛 BUG: Props are not inferred with a Svelte generic component
### Describe the Bug

Before all, once again, I'm note sure if it is a compiler or a language-tools issue. I guess it is not the compiler since the app can be build without errors. Sorry if I picked up the wrong repository.

## Context

Svelte supports the creation of [generic components using Typescript](https://github.com/dummdidumm/rfcs/blob/ts-typedefs-within-svelte-components/text/ts-typing-props-slots-events.md#generics). The syntax that was retained looks like:

```svelte
<script lang="ts" generics="T extends boolean, X">
    import {createEventDispatcher} from "svelte";

    export let array1: T[];
    export let item1: T;
    export let array2: X[];

    const dispatch = createEventDispatcher<{arrayItemClick: X}>();
</script>
```

However, it seems this API is [still experimental](https://svelte.dev/docs/typescript#experimental-advanced-typings) and not yet documented on the Svelte website.

## The issue

When importing a generic component in a Svelte file, everything works as expected: the props are correctly inferred and we can have auto-completion in VS Code.

When importing the same generic component in an Astro file, the app can be build (no errors, and the component is visible in the browser). However, the Props are not inferred and Typescript complains.

## Example

With this component:

```svelte
<script lang="ts" generics="T extends boolean = false">
  import type { SvelteHTMLElements } from "svelte/elements";

  type BaseProps = T extends true
    ? SvelteHTMLElements["ol"]
    : SvelteHTMLElements["ul"];

  type $$Props = BaseProps & {
    isOrdered?: T;
  };

  export let isOrdered = false;

  let tag: Extract<keyof HTMLElementTagNameMap, "ol" | "ul">;
  $: tag = isOrdered ? "ol" : "ul";
</script>

<svelte:element this={tag} {...$$restProps}>
  <slot />
</svelte:element>
```

If I try to import it in a file with the `.astro` extension, I can't benefit from the intellisense.

import of a Svelte generic component in Astro

And Typescript gives me the following error:

```
Type '{}' is not assignable to type 'IntrinsicAttributes & ComponentConstructorOptions<$$Props>'.
  Property 'target' is missing in type '{}' but required in type 'ComponentConstructorOptions<$$Props>'.ts(2322)
```

## My expectations

No errors from Typescript and the ability to use the intellisense to fill the props, like any component.

## System info

```
> astro info

Astro                    v4.3.2
Node                     v18.17.1
System                   Linux (x64)
Package Manager          npm
Output                   static
Adapter                  none
Integrations             @astrojs/svelte
which: no xclip in .......
```

### Steps to Reproduce

I created [a Stackblitz](https://stackblitz.com/edit/astro-svelte-typescript-generics?file=src%2Fcomponents%2FAstroTest.astro) that can be downloaded and opened in VS Code (since the parser won't work otherwise). It contains three files in `src/components`:
* a generic Svelte component (`List.svelte`)
* a Svelte file to test the import (`SvelteTest.svelte`)
* an Astro file to test the import (`AstroTest.svelte`)

Here the steps to manually reproduce the error:

1. In VS Code, create a new Astro project with `npm create astro@latest -- --template framework-svelte` (and install the Astro and Svelte extensions if needed)
2. Answer `Yes` for everything (and keep `Strict` for Typescript)
3. Create a file `src/components/List.svelte` that contains our generic component:

```svelte
<script lang="ts" generics="T extends boolean = false">
  import type { SvelteHTMLElements } from "svelte/elements";

  type BaseProps = T extends true
    ? SvelteHTMLElements["ol"]
    : SvelteHTMLElements["ul"];

  type $$Props = BaseProps & {
    isOrdered?: T;
  };

  export let isOrdered = false;

  let tag: Extract<keyof HTMLElementTagNameMap, "ol" | "ul">;
  $: tag = isOrdered ? "ol" : "ul";
</script>

<svelte:element this={tag} {...$$restProps}>
  <slot />
</svelte:element>
```

4. Import this component in another Svelte file `src/components/SvelteTest.svelte`:

```svelte
<script>
  import List from "./List.svelte";
</script>

<List />
```

5. Import the component in an Astro file `src/components/AstroTest.astro`:

```astro
---
import ListSvelte from "./List.svelte";
---

<ListSvelte />
```
6. Now if we look our two tests files:
* the Svelte one correctly infer the Props: `class List<T extends boolean = false>`
* the Astro one does not infer the Props and Typescript is complaining with:

```
Type '{}' is not assignable to type 'IntrinsicAttributes & ComponentConstructorOptions<$$Props>'.
  Property 'target' is missing in type '{}' but required in type 'ComponentConstructorOptions<$$Props>'.ts(2322)
```

### PR DESCRIPTION
**Pull request by seroperson**
Closes #14669

## Changes

When a Svelte component uses generic type parameters (`<script generics="T extends ..."`>), the editor integration now preserves the generics in the generated TypeScript, so props are correctly inferred and intellisense works in `.astro` files.

## Testing

Added several astro check tests in `svelte/test/check.test.js`

- Single-param generic with conditional type constraint
- Multi-param generic with nested angle brackets (`Record<string, unknown>`)
- Arrow function type in generic constraint (`(val: string) => boolean`)

To check this manually:

1. Create a Svelte component with `generics` attribute:
```svelte
<script lang="ts" generics="T extends boolean = false">
  interface Props { flag?: T; items: T extends true ? number[] : string[] }
  let { flag, items }: Props = $props();
</script>
```
2. Import and use it in an `.astro` file
2. Before the fix: TypeScript errors and no intellisense
2. After the fix: props are correctly inferred with generic constraints

## Docs

No docs changes are required.

### CODE REVIEW

File: packages/integrations/svelte/src/editor.cts [hunk 1]
```diff
@@ -29,3 +38,39 @@ export function toTSX(code: string, className: string): string {
 
 	return result;
 }
+
+function extractGenerics(tsx: string): { params: string; names: string } | null {
```
Comment (ematipico) [inline packages/integrations/svelte/src/editor.cts]: You could export this function and add some unit tests

Comment (seroperson) [reply packages/integrations/svelte/src/editor.cts]: Done. Also improved this function just a little to cover some possible edge cases. Looks weird, but works.

**Comment by seroperson** [comment]
VSCode e2e test just for the history

image
